runtime/vm/mz64/UpcallThunkGen.cpp

```diff
@@ -56,179 +57,433 @@ extern "C" {
+	for (I_32 i = 0; i < lastSigIdx; i++) {
+		U_32 argSizeInBytes = sigArray[i].sizeInByte;
+		switch (sigArray[i].type) {
+		case J9_FFI_UPCALL_SIG_TYPE_CHAR:    /* Fall through */
+		case J9_FFI_UPCALL_SIG_TYPE_SHORT:   /* Fall through */
+		case J9_FFI_UPCALL_SIG_TYPE_INT32:   /* Fall through */
+		case J9_FFI_UPCALL_SIG_TYPE_POINTER: /* Fall through */
+		case J9_FFI_UPCALL_SIG_TYPE_INT64:
+			if (gprArgCount < maxGPRArgCount) {
+				numInstructions++;
+				gprArgCount++;
+			}
+			break;
+		case J9_FFI_UPCALL_SIG_TYPE_FLOAT:
+		case J9_FFI_UPCALL_SIG_TYPE_DOUBLE:
+			if (fprArgCount < maxFPRArgCount) {
+				numInstructions++;
+				fprArgCount++;
+			}
+			if (i < 3) {
+				gprArgCount++; // If one of the first three doublewords of the argument area is float/double, then the corresponding GPR is undefined. So we must ignore any value in there.
+			}
+			break;
+		case J9_FFI_UPCALL_SIG_TYPE_STRUCT_AGGREGATE_ALL_SP:
+			if (argSizeInBytes == 2 * sizeof(float)) {
+				// a pure float struct with only 2 floats is passed in 2 FPRs.
+				// If only one FPR is free, then the first half of the struct is
+				// in FPR6, and the second half is in the corresponding slot
+				// in the argument area, right adjusted.
+				if (fprArgCount <= maxFPRArgCount - 2) {
+					numInstructions += 2;
+					fprArgCount += 2;
+				}
+				else if (fprArgCount < maxFPRArgCount) {
+					numInstructions++;
+					fprArgCount++;
+				}
+			}
+			else { // All other pure float structs are passed in GPR (if there's space)
+				if (gprArgCount < maxGPRArgCount) {
+					// gprsNeededForEntireStruct below represents the number of GPRs we would need hypothetically to load the entire struct in GPRs.
+					// Since this is a pure float structure, the size will always be a multiple of 4. So if we divide by 8, like below, the result
+					// will either be some whole number 'X', or 'X.5'. For `X.5`, adding 0.5 to the sum will always round UP to the nearest whole number.
+					int gprsNeededForEntireStruct = (int)(((float)argSizeInBytes / 8.0F) + 0.5);
 
-    }
+					// We may not have all the GPRs available to store the entire struct. In that case, we store only what's in GPR (the rest of the struct will already be in memory.)
+					int numInstructionsAvailable = maxGPRArgCount - gprArgCount;
+					numInstructions += std::min(gprsNeededForEntireStruct, numInstructionsAvailable);
+					gprArgCount += std::min(gprsNeededForEntireStruct, numInstructionsAvailable);
+				}
+			}
+			break;
+		case J9_FFI_UPCALL_SIG_TYPE_STRUCT_AGGREGATE_ALL_DP:
+			if (argSizeInBytes == 2 * sizeof(double)) {
+				// a pure double struct with only 2 floats is passed in 2 FPRs.
+				// If only one FPR is free, then the first half of the struct is
+				// in FPR6, and the second half is in the corresponding memory slot
+				// in the stack's argument area.
+				if (fprArgCount <= maxFPRArgCount - 2) {
+					numInstructions += 2;
+					fprArgCount += 2;
```

Comment:
Similar to question I asked couple of lines back, don't we need to update gprCount as well?

Reply:
This, along with the other places where gprArgCount needs to be incremented (and not incremented) are fixed in: 53c52c9835f66d28f5b4b3e0120ae281ed8f0b02